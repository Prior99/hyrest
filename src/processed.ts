export interface ProcessedInput<T> {
    value?: T;
    nested?: {
        [key: number]: Processed<any> | Processed<any>[];
        [key: string]: Processed<any> | Processed<any>[];
    };
    errors?: string[];
}

export interface MergeOptions {
    skipValue?: boolean;
}

/**
 * Represents the result of a full validation of a value, array or schema.
 */
export class Processed<T> {
    constructor(copy?: ProcessedInput<T>) {
        Object.assign(this, copy);
    }

    /**
     * The parsed value, only set if no errors occured.
     */
    public value?: T;
    /**
     * A map with the results of nested objects. On a schema,
     * for each key with an error a key would be present in this map.
     */
    public nested?: { [key: string]: Processed<any>; };
    /**
     * A list of all errors occured for this value.
     */
    public errors?: string[];

    /**
     * Adds the errors to the list of errors, initializing them if necessary.
     *
     * @param errors The list of errors that occured.
     */
    public addErrors (...errors: string[]) {
        if (errors.length === 0) {
            return;
        }
        if (!this.errors) {
            this.errors = [...errors];
            return;
        }
        this.errors.push(...errors);
    }

    /**
     * Adds the validation result for a nested property to the map of nested properties.
     * Initializies the map if necessary.
     *
     * @param key The name of the property for which the validation result should be set.
     * @param nested The result of the validation for the given property.
     */
    public addNested (key: string | number, nested: Processed<any>) {
        if (!this.nested) { this.nested = { [key]: nested }; }
        this.nested[key] = nested;
    }

    /**
     * Will be `true` if any errors occured for this value or nested ones.
     * Recursively traverses all nested values.
     */
    public get hasErrors(): boolean {
        const { errors, nested } = this;
        const currentHasErrors = Boolean(errors && errors.length > 0);
        const nestedHasErrors: boolean = Boolean(nested && Object.keys(nested).reduce((result, key) => {
            return result || nested[key].hasErrors;
        }, false));
        return  currentHasErrors || nestedHasErrors;
    }

    /**
     * Merges the results (errors, value and nested) of another validation result with this one.
     * This will mutate this instance. This function will recursively traverse all nested results.
     *
     * @param other The other value to merge into this one.
     */
    public merge(other: Processed<T>, options?: MergeOptions) {
        if (typeof other === "undefined" || other === null) {
            return;
        }
        // Only merge errors if the other has any errors.
        if (typeof other.errors !== "undefined") {
            // Initialize our `errors` if it wasn't initialized previously.
            if (this.errors) {
                this.errors.push(...other.errors);
            } else {
                this.errors = [...other.errors];
            }
        }
        // If set, merge all nested values.
        if (typeof other.nested !== "undefined" && other.nested !== null) {
            Object.keys(other.nested).forEach(key => {
                if (typeof this.nested === "undefined" || typeof this.nested[key] === "undefined") {
                    this.addNested(key, other.nested[key]);
                    return;
                }
                // Merge nested values recursively.
                this.nested[key].merge(other.nested[key]);
            });
        }
        const skipValue = Boolean(options && options.skipValue);
        // Copy the value if our value is not set but the other value is set.
        if (typeof this.value === "undefined" && typeof other.value !== "undefined" && !skipValue) {
            this.value = other.value;
        }
    }
}
