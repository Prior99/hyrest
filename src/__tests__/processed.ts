import { Processed } from "../processed";

test("Adding errors to `Processed`", () => {
    const processed = new Processed();
    processed.addErrors("test 1", "test 2", "test 3");
    expect(processed).toMatchSnapshot();
    processed.addErrors("test 4", "test 5");
    expect(processed).toMatchSnapshot();
});

test("Adding no errors to `Processed`", () => {
    const processed = new Processed();
    processed.addErrors();
    expect(processed).toMatchSnapshot();
});

test("Adding nested results to processed", () => {
    const processed = new Processed();
    processed.addNested("a", new Processed({ errors: ["test"] }));
    expect(processed).toMatchSnapshot();
    processed.addNested("b", new Processed({ errors: ["test"] }));
    expect(processed).toMatchSnapshot();
});

test("`Processed.hasErrors`", () => {
    const processed = new Processed();
    expect(processed.hasErrors).toBe(false);
    processed.addErrors("test");
    expect(processed.hasErrors).toBe(true);
});

test("`Processed.hasErrors` with nested results", () => {
    const processed = new Processed();
    const otherProcessed = new Processed();
    otherProcessed.addNested("b", new Processed({ errors: ["error"] }));
    processed.addNested("a", otherProcessed);
    expect(processed.hasErrors).toBe(true);
});

test("`Processed.merge`", () => {
    const processed = new Processed();
    processed.merge(new Processed({ value: "test" }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({ errors: ["test"] }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({
        nested: {
            a: new Processed({ errors: ["test"] }),
        },
    }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({ value: "invalid" }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({ errors: ["another test"] }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({
        nested: {
            a: new Processed({ errors: ["a second test"] }),
            b: new Processed({
                nested: {
                    c: new Processed({ errors: ["deeply nested"] }),
                },
            }),
        },
    }));
    expect(processed).toMatchSnapshot();
    processed.merge(new Processed({
        nested: {
            b: new Processed({
                nested: {
                    c: new Processed({ errors: ["another deeply nested one"] }),
                },
            }),
        },
    }));
    expect(processed).toMatchSnapshot();
    processed.merge(undefined);
    expect(processed).toMatchSnapshot();
});
