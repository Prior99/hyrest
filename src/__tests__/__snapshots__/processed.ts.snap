// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`Processed.merge\` 1`] = `
Processed {
  "value": "test",
}
`;

exports[`\`Processed.merge\` 2`] = `
Processed {
  "errors": Array [
    "test",
  ],
  "value": "test",
}
`;

exports[`\`Processed.merge\` 3`] = `
Processed {
  "errors": Array [
    "test",
  ],
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
      ],
    },
  },
  "value": "test",
}
`;

exports[`\`Processed.merge\` 4`] = `
Processed {
  "errors": Array [
    "test",
  ],
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
      ],
    },
  },
  "value": "test",
}
`;

exports[`\`Processed.merge\` 5`] = `
Processed {
  "errors": Array [
    "test",
    "another test",
  ],
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
      ],
    },
  },
  "value": "test",
}
`;

exports[`\`Processed.merge\` 6`] = `
Processed {
  "errors": Array [
    "test",
    "another test",
  ],
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
        "a second test",
      ],
    },
    "b": Processed {
      "nested": Object {
        "c": Processed {
          "errors": Array [
            "deeply nested",
          ],
        },
      },
    },
  },
  "value": "test",
}
`;

exports[`\`Processed.merge\` 7`] = `
Processed {
  "errors": Array [
    "test",
    "another test",
  ],
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
        "a second test",
      ],
    },
    "b": Processed {
      "nested": Object {
        "c": Processed {
          "errors": Array [
            "deeply nested",
            "another deeply nested one",
          ],
        },
      },
    },
  },
  "value": "test",
}
`;

exports[`Adding errors to \`Processed\` 1`] = `
Processed {
  "errors": Array [
    "test 1",
    "test 2",
    "test 3",
  ],
}
`;

exports[`Adding errors to \`Processed\` 2`] = `
Processed {
  "errors": Array [
    "test 1",
    "test 2",
    "test 3",
    "test 4",
    "test 5",
  ],
}
`;

exports[`Adding nested results to processed 1`] = `
Processed {
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
      ],
    },
  },
}
`;

exports[`Adding nested results to processed 2`] = `
Processed {
  "nested": Object {
    "a": Processed {
      "errors": Array [
        "test",
      ],
    },
    "b": Processed {
      "errors": Array [
        "test",
      ],
    },
  },
}
`;

exports[`Adding no errors to \`Processed\` 1`] = `Processed {}`;
